model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java

```diff
@@ -87,43 +81,47 @@ public GroupModel getGroupById(RealmModel realm, String id) {
             return null;
         }
         
-        MapGroupEntity entity = tx.get(uid, groupStore::get);
-        return (entity == null || ! entityRealmFilter(realm).test(entity))
+        MapGroupEntity entity = tx.read(uid);
+        String realmId = realm.getId();
+        return (entity == null || ! Objects.equals(realmId, entity.getRealmId()))
                 ? null
                 : entityToAdapterFunc(realm).apply(entity);
     }
 
-    private Stream<MapGroupEntity> getNotRemovedUpdatedGroupsStream() {
-        Stream<MapGroupEntity> updatedAndNotRemovedGroupsStream = groupStore.entrySet().stream()
-                .map(tx::getUpdated)    // If the group has been removed, tx.get will return null, otherwise it will return me.getValue()
-                .filter(Objects::nonNull);
-        return Stream.concat(tx.createdValuesStream(), updatedAndNotRemovedGroupsStream);
-    }
-
-    private Stream<MapGroupEntity> getUnsortedGroupEntitiesStream(RealmModel realm) {
-        return getNotRemovedUpdatedGroupsStream()
-                .filter(entityRealmFilter(realm));
-    }
-
     @Override
     public Stream<GroupModel> getGroupsStream(RealmModel realm) {
+        return getGroupsStreamInternal(realm, null);
+    }
+
+    private Stream<GroupModel> getGroupsStreamInternal(RealmModel realm, UnaryOperator<ModelCriteriaBuilder<GroupModel>> modifier) {
         LOG.tracef("getGroupsStream(%s)%s", realm, getShortStackTrace());
-        return getUnsortedGroupEntitiesStream(realm)
+        ModelCriteriaBuilder<GroupModel> mcb = groupStore.createCriteriaBuilder()
+          .compare(SearchableFields.REALM_ID, Operator.EQ, realm.getId());
```

Comment:
I was thinking about it, I decided for a more explicit "WHERE clause" as it then shows exact condition in full and is still short enough. I'm open to change it though if there is a convincing argument though

Reply:
I see. Well I don't have any strong argument there, I just think that it would be worth adding as almost each entity we have is related to exactly one realm. But I agree with you, the code is more comprehensible this way, so we can leave it as is.